app/src/main/java/com/PrivacyGuard/Application/Fragments/LeakQueryFragment.java

```diff
@@ -140,95 +151,122 @@ public void onClick(View v) {
+                leaksList.setAdapter(new ListAdapter(getContext(), new ArrayList<DataLeak>()));
+                totalNumber.setVisibility(View.GONE);
+                progressBar.setVisibility(View.VISIBLE);
+                query.setEnabled(false);
+                query.setAlpha(0.3f);
+
+                new LoadQueryData().execute(spinnerCategory.getSelectedItem().toString(),
+                        spinnerStatus.getSelectedItem().toString());
+            }
+        });
+
+        return view;
+    }
+
+    private class ListAdapter extends ArrayAdapter<DataLeak> {
+        public ListAdapter(Context context, ArrayList<DataLeak> leaks) {
+            super(context, 0, leaks);
+        }
+
+        @Override
+        public View getView(int position, View convertView, ViewGroup parent) {
+            DataLeak dataLeak = getItem(position);
+            // Check if an existing view is being reused, otherwise inflate the view
+            if (convertView == null) {
+                convertView = LayoutInflater.from(getContext()).inflate(R.layout.data_leak, parent, false);
+            }
+
+            TextView appNameText = (TextView)convertView.findViewById(R.id.app_name);
+            TextView categoryText = (TextView)convertView.findViewById(R.id.category);
+            TextView typeText = (TextView)convertView.findViewById(R.id.type);
+            TextView timeStampText = (TextView)convertView.findViewById(R.id.time_stamp);
+            TextView contentText = (TextView)convertView.findViewById(R.id.content);
+            TextView statusText = (TextView)convertView.findViewById(R.id.status);
+
+            String categoryCamelCase = dataLeak.getCategory().toLowerCase();
+            categoryCamelCase = categoryCamelCase.substring(0,1).toUpperCase() + categoryCamelCase.substring(1);
+
+            appNameText.setText(dataLeak.getAppName());
+            categoryText.setText(categoryCamelCase);
+            typeText.setText(dataLeak.getType());
+            timeStampText.setText(dateFormatDisplaySpecific.format(dataLeak.getTimestampDate()));
+            contentText.setText(dataLeak.getLeakContent());
+            statusText.setText(dataLeak.getForegroundStatus() == DatabaseHandler.FOREGROUND_STATUS ? FOREGROUND : BACKGROUND);
+
+            return convertView;
+        }
+    }
 
-                List<DataLeak> revisedLeaks = new ArrayList<>();
+    private class LoadQueryData extends AsyncTask<String, Void, ArrayList<DataLeak>> {
 
-                for (DataLeak dataLeak : leaks) {
-                    if (status.equals(foreground) && dataLeak.getForegroundStatus() != DatabaseHandler.FOREGROUND_STATUS) {
-                        continue;
-                    }
+        @Override
+        protected ArrayList<DataLeak> doInBackground(String... params) {
+            List<DataLeak> leaks = new ArrayList<>();
+            String category = params[0];
+            String status = params[1];
+
+            if (category.equals(ALL)) {
```

Comment:
category and status should never be null.  If either one is null, then a null pointer exception would alert a programmer that they are using the AsyncTask incorrectly.